docs/src/design/multiset.md

```diff
@@ -1,56 +1,158 @@
+
+These specialized components are designed to prove the internal correctness of the execution of the operations they support. However, in isolation they cannot make any guarantees about the source of the input data or the destination of the output data.
+
+In order to prove that the inputs and outputs specified by the main circuit match the inputs and outputs provably executed in the specialized component, some kind of provable communication bus is needed. 
+
+This bus is typically implemented as some kind of lookup argument, and in Miden VM in particular we use multiset checks.
+
+### Implementation
+
+A `bus` can be implemented as a single trace column $b$ where a request can be sent to a specific component and a corresponding response will be sent back by that component. 
+
+The values in this column contain a running product of the communication with the component as follows:
+
+- Each request is “sent” by computing a lookup value from some information that's specific to the specialized component, the operation inputs, and the operation outputs, and then dividing it out of the running product column $b$.
+- Each chiplet response is “sent” by computing the same lookup value from the component-specific information, inputs, and outputs, and then multiplying it into the running product column $b$.
+
+Thus, if the requests and responses match, and the bus column $b$ is initialized to $1$, then $b$ will start and end with the value $1$. This condition is enforced by boundary constraints on column $b$.
+
+Note that the order of the requests and responses does not matter, as long as they are all included in $b$. In fact, requests and responses for the same operation will generally occur at different cycles. Additionally, there could be multiple requests sent in the same cycle, and there could also be a response provided at the same cycle that a request is received.
+
+### Communication bus constraints
+
+These constraints can be expressed in a general way with the 2 following requirements:
+
+- The lookup value must be computed using random values $\alpha_0, \alpha_1$, etc. that are provided by the verifier after the prover has committed to the main execution trace. 
+- The lookup value must include all uniquely identifying information for the component/operation and its inputs and outputs.
+
+Given an operation $op_{example}$ with inputs $i_0, i_1, ..., i_n$ and outputs $o_0, o_1, ..., o_m$, the lookup value can be computed as follows:
+
+$$lookup = \alpha_0 + \alpha_1 \cdot op_{example} + \alpha_2 \cdot i_0 + \alpha_3 \cdot i_1 + ... + \alpha_{n+2} \cdot i_n + \alpha_{n+3} \cdot o_0 + \alpha_{n+4} \cdot o_1 + ... + \alpha_{n + 2 + m} \cdot o_m$$
```

Comment:
This formula overflows the page for me. Maybe we can shorten it a bit. For example, use $op_{ex}$ instead of $op_{example}$, or get rid of the terms for $i_1$ and $o_1$?

Reply:
I'll try these in the next PR